src/tree/tree_model.cc

```diff
@@ -169,4 +169,236 @@ std::string RegTree::DumpModel(const FeatureMap& fmap,
   }
   return fo.str();
 }
+void RegTree::FillNodeMeanValues() {
+  size_t num_nodes = this->param.num_nodes;
+  if (this->node_mean_values_.size() == num_nodes) {
+    return;
+  }
+  this->node_mean_values_.resize(num_nodes);
+  for (int root_id = 0; root_id < param.num_roots; ++root_id) {
+    this->FillNodeMeanValue(root_id);
+  }
+}
+
+bst_float RegTree::FillNodeMeanValue(int nid) {
+  bst_float result;
+  auto& node = (*this)[nid];
+  if (node.IsLeaf()) {
+    result = node.LeafValue();
+  } else {
+    result  = this->FillNodeMeanValue(node.LeftChild()) * this->Stat(node.LeftChild()).sum_hess;
+    result += this->FillNodeMeanValue(node.RightChild()) * this->Stat(node.RightChild()).sum_hess;
+    result /= this->Stat(nid).sum_hess;
+  }
+  this->node_mean_values_[nid] = result;
+  return result;
+}
+
+void RegTree::CalculateContributionsApprox(const RegTree::FVec& feat, unsigned root_id,
+                                                  bst_float *out_contribs) const {
```

Comment:
This line is incorrectly indented.  I don't know how it looks like on your IDE, but please view it on Github.